File: leetcode/word-search-ii.cpp
// Author: Suraj Aralihalli
// Url: https://leetcode.com/problems/word-search-ii/
// Date: 27th August, 2022
// Tags: trie, backtracking, dfs


//Approach 2
class Solution {
public:
    vector<vector<char>> mat;
    int n;
    int m;
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        n = board.size();
        m = board[0].size();
        mat = board;
        vector<vector<int>> marked(n, vector<int>(m, 0));
        vector<string> wordsFound;
        for(auto word: words)
        {
            bool wordFound = false;
            for(int i=0;i<n;i++)
            {
                for(int j=0;j<m;j++)
                {
                    if(board[i][j]==word[0])
                    {
                        marked[i][j]=1;
                        wordFound = findWord(marked, word, 1, i, j);
                        marked[i][j]=0;
                    }
                    
                    if(wordFound) 
                    {
                        wordsFound.push_back(word);
                        break;
                    }
                }
                if(wordFound) break;
            }   
        }
        
        return wordsFound;
        
    }
    
    bool findWord(vector<vector<int>>& marked, string word, int index, int x, int y)
    {
        int l = word.size();
        if(index==l) return true;
        else
        {
            bool wordFound = false;
            if(!wordFound && x+1<n && marked[x+1][y]==0 &&  mat[x+1][y]==word[index])
            {
                
                marked[x+1][y]=1;
                wordFound = findWord(marked, word, index+1, x+1, y);
                marked[x+1][y]=0;
            }
            if(!wordFound && x-1>=0 && marked[x-1][y]==0 && mat[x-1][y]==word[index])
            {
                marked[x-1][y]=1;
                wordFound = findWord(marked, word, index+1, x-1, y);
                marked[x-1][y]=0;
            }
            if(!wordFound && y+1<m && marked[x][y+1]==0 && mat[x][y+1]==word[index])
            {
                marked[x][y+1]=1;
                wordFound = findWord(marked, word, index+1, x, y+1);
                marked[x][y+1]=0;
            }
            if(!wordFound && y-1>=0 && marked[x][y-1]==0 && mat[x][y-1]==word[index])
            {
                marked[x][y-1]=1;
                wordFound = findWord(marked, word, index+1, x, y-1);
                marked[x][y-1]=0;
                
            }
            
            return wordFound;
            
        }
    }
    
    
};

File: leetcode/word-search-ii.py
// Author: Suraj Aralihalli
// Url: https://leetcode.com/problems/word-search-ii/
// Date: 28th August, 2022
// Tags: trie, prefix-tree

class TrieNode:
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.isWord = False
        
class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word:str):
        node = self.root
        for c in word:
            node = node.children[c]
        node.isWord = True

    def search(self, word:str) -> bool:
        node = self.root
        for c in word:
            node = node.children.get(c)
            if not node:
                return False
        return True
        
class Solution:
    def __init__(self):
        self.mat = None
        self.sol = []
        
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        trie = Trie()
        
        for word in words:
            trie.insert(word)
            
        node = trie.root
        self.mat = board
        for i in range(len(board)):
            for j in range(len(board[0])):
                self.dfs(node, i, j, "")
        
        return self.sol
        
    def dfs(self, node: TrieNode, i:int, j:int, path:""):
        
        if(node.isWord):
            self.sol.append(path)
            node.isWord = False
            
        if(i >= len(self.mat) or i < 0 or j >= len(self.mat[0]) or j < 0):
            return
        
        c = self.mat[i][j]
        if(node.children.get(c) == None):
            return
        else:
            nextNode = node.children.get(c)
            self.mat[i][j] = "#"
            self.dfs(nextNode, i+1, j, path+c)
            self.dfs(nextNode, i-1, j, path+c)
            self.dfs(nextNode, i, j+1, path+c)
            self.dfs(nextNode, i, j-1, path+c)
            self.mat[i][j] = c
            return
